### ISSUE TITLE
Adding handle to data store in detached container doesn't bind the data store

### ISSUE THREAD

**Issue by agarwal-navin** [microsoft/FluidFramework#9127]
This bug prevents the removal of `bindToContext` API that is called on a data store after its creation. We need to remove this asap as that affects GC.

In detached container, when a handle to a data store is attached to a bound DDS (DDS in root data stores), the data store is not bound. So, requesting it via the `request` API hangs forever. See the test added by this PR for an example of this - https://github.com/microsoft/FluidFramework/pull/9129. This test should be enabled once this bug is fixed.

Expected behavior - When a data store's handle is added to a bound DDS, it should be reachable and returned via the `request` APIs irrespective of whether the container is attached or not.

**Comment by agarwal-navin** [comment]
The bug here is that in detached container, root data stores and their DDSs have their `graphAttachState` as `Detached`. Their `bindState` is however `Bound`.
When a non-root data store handle is stored in a root data store, it is not bound because the `graphAttachState` is detached. Essentially, nothing in a detached container has their `graphAttachState` as attached.

The fix for this can be one of the following:
1. Switch the `graphAttachState` to `Attached` for root data stores. I had tried this before and ran into issues. See point # 4 in https://github.com/microsoft/FluidFramework/issues/8347 for details.
2. Separate bind state and graph attach state. When a handle is added to a DDS, if the DDS is bound, bind the handle as well and if the DDS's graph is attached, attach the handle's graph as well. We might run into similar issues as I mentioned in # 1 above so we have to be careful. Also, this won't fix the underlying problem of data store's `graphAttachState` not being attached but we can do that separately.
3. When a data store is requested via the request / resolve APIs, return it whether it is bound or not. This doesn't fix the underlying problem but it will unblock the integration of release/0.57 where `bindToContext` is already removed. The issue with this approach is that it will open up another path to retrieve data stores via the `request` APIs when we are trying to get rid of `request` patterns altogether

**Comment by anthony-murphy** [comment]
I never understood the difference between bindState, and graphAttachState, I think @jatgarg understands it. It always seemed redundant to me, as we should only really need attach and bind. Anything unbound is detached, anything bound to anything attached is also attached. On attach, all bound things to that attached thing are also attached.

**Comment by anthony-murphy** [comment]
> When a non-root data store handle is stored in a root data store, it is not bound because the `graphAttachState` is detached. Essentially, nothing in a detached container has their `graphAttachState` as attached.

looking at the code, this seems wrong. graphAttachState should only matter during the attach process. where is this in the code?

**Comment by agarwal-navin** [comment]
> looking at the code, this seems wrong. graphAttachState should only matter during the attach process. where is this in the code?

In the handle's bind method -https://github.com/microsoft/FluidFramework/blob/03a57776939a5cab5305b62c275046b417b1832f/packages/runtime/datastore/src/fluidHandle.ts#L65

**Comment by anthony-murphy** [comment]
if not attached or graph attach state is not detached. in a detached container attach is false, and graph attach state should be detached, so this condition should be skipped, and the datastore is just bound.

I think the problem with that code is that when the container is attached or attaching, and the graph attach state is attached or attaching we return, so we never bind it, which seem wrong. it seems like we should always bind

**Comment by agarwal-navin** [comment]
This is super confusing because of 3 different state tracking :-). But I think we are on the same page regarding how this should work.

My understanding is that graphAttachState exists only to break cycles when attaching an object. So, it should not be used to make decision of whether to bind an object or not.
The way it should probably work is - if a handle is added to a bound object, it should also be bound. It should not care about the attach state or graph attach state. And it has to do this by calling `bindToContext` and not `attachGraph`.
I had tried to always call `attachGraph` (https://github.com/microsoft/FluidFramework/pull/7917) when binding an object and it let to [this bug](https://github.com/microsoft/FluidFramework/issues/8305). While I would still like to resurrect the above PR, we will have to fix the bug and be very careful.

**Comment by agarwal-navin** [comment]
Based on offline discussions, the path forward for this is:
1. Remove `graphAttachState` and use `bindState` to determine whether a handle that is added to a data store should be bound or not.
2. Basically, do what #7917 did but remove `graphAttachState` instead of `bindState`. Additionally, when a data store is attached , explicitly call `markAttached()` on the `LocalChannelContext` for channels that are not yet attached. This will prevent the following bug that happened when #7917 was merged - https://github.com/microsoft/FluidFramework/issues/8305.
3. There are couple tests specific to FFX scenarios that I added to ensure we don't break them - https://github.com/microsoft/FluidFramework/pull/9158 and https://github.com/microsoft/FluidFramework/pull/9129. https://github.com/microsoft/FluidFramework/pull/9129 is currently disabled and should be enabled with this fix and should work.
4. There are these attach life cycle tests that should also validate scenarios around attachment - https://github.com/microsoft/FluidFramework/blob/32631566ef262c90edb385beebf60215672b0a78/packages/test/test-end-to-end-tests/src/test/attachLifecycle.spec.ts
